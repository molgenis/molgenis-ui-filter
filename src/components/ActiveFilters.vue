<template>
  <div>
    <button
      v-for="(item, key) in activeValues"
      :key="key"
      type="button"
      class="active-filter btn btn-light m-1 btn-outline-secondary"
      @click="removeFilter(item)"
    >
      {{ item.label }}: {{ item.value }}
      <font-awesome-icon
        icon="times"
        class="ml-1"
      />
    </button>
  </div>
</template>

<script>
import Vue from 'vue'

import { library } from '@fortawesome/fontawesome-svg-core'
import { faTimes } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faTimes)

export default Vue.extend({
  name: 'ActiveFilters',
  components: { FontAwesomeIcon },
  props: {
    filters: {
      type: Array,
      required: true
    },
    value: {
      type: Object,
      default: () => ({})
    }
  },
  data () {
    return {
      activeValues: []
    }
  },
  watch: {
    value: {
      handler (newValue) {
        this.buildActiveValues(newValue)
      },
      immediate: true
    }
  },
  methods: {
    async buildActiveValues (newValue) {
      const activeValues = []
      Object.entries(newValue).forEach(async ([key, current]) => {
        const filter = this.selectFilter(key)

        // Clean op the values by removing undefined entry's
        if (current === undefined || !current.length) return

        // Unpack array
        if (Array.isArray(current)) {
          // Checkbox
          if (filter.type === 'checkbox-filter') {
            // resolve options function/promise and show results later
            const option = await filter.options()
            current.forEach(subKey => {
              const findTextFromValue = option.filter(
                filter => filter.value === subKey
              )[0]
              activeValues.push({
                key,
                subKey,
                value: findTextFromValue.text,
                label: filter.label
              })
            })
          }
          // Range Filter
          if (filter.type === 'range-filter') {
            activeValues.push({
              key,
              value: `${current[0]} to ${current[1]}`,
              label: filter.label
            })
          }
          if (filter.type === 'multi-filter') {
            const options = await filter.options(`?q=id=in=(${current.join(',')})`)
            current.forEach(subKey => {
              const findTextFromValue = options.filter(
                filter => filter.value === subKey
              )[0]
              activeValues.push({
                key,
                subKey,
                value: findTextFromValue.text,
                label: filter.label
              })
            })
          }
        } else {
          // Single item
          activeValues.push({ key, value: current, label: filter.label })
        }
      })
      if (this.value === newValue) {
        this.activeValues = activeValues
      }
    },
    selectFilter (key) {
      return this.filters.filter(filter => filter.name === key)[0]
    },
    removeFilter ({ key, subKey }) {
      if (subKey === undefined) {
        let selections = { ...this.value }
        delete selections[key]
        this.$emit('input', selections)
      } else {
        let selections = { ...this.value }
        selections[key] = selections[key].filter(key => key !== subKey)
        this.$emit('input', selections)
      }
    }
  }
})
</script>

<style scoped>
button svg path {
  transition: fill 0.3s;
}
button:hover svg path {
  fill: var(--danger);
}
</style>
