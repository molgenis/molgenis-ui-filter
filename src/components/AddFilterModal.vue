<template>
  <div>
    <b-button
      v-b-modal.modal-add-filter
      variant="outline-secondary"
      class="w-100 mt-2 add-button"
    >
      Add a filter
      <font-awesome-icon
        class="ml-1"
        icon="plus"
      />
    </b-button>

    <b-modal
      id="modal-add-filter"
      title="Add filter"
      @ok="addFilter"
      @show="selected = inactiveFilters[0].id"
    >
      <b-form-select
        v-model="selected"
        name="filter"
        :options="inactiveFilters"
        value-field="id"
        text-field="name"
      />
    </b-modal>
  </div>
</template>

<script>
import Vue from 'vue'

import { library } from '@fortawesome/fontawesome-svg-core'
import { faPlus } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faPlus)

export default Vue.extend({
  name: 'AddFilterModal',
  components: { FontAwesomeIcon },
  store: {
    'filters': 'filters.available'
  },
  data () {
    return {
      selected: null
    }
  },
  computed: {
    inactiveFilters () {
      return this.filters.filter(filter => !filter.active).map((f) => ({
        id: f.id,
        name: f.label
      }))
    }
  },
  methods: {
    addFilter () {
      const filter = this.filters.find((f) => f.id === this.selected)
      filter.active = true
    }
  }
})
</script>
