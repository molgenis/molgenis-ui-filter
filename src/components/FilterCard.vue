<template>
  <b-card
    no-body
    class="filter-card mb-2"
  >
    <b-card-header
      :class="filter.collapsible ? 'collapsable' : ''"
      @click="toggleCollapse"
    >
      <div
        class="title mr-3 px-1"
        :title="filter.label"
      >
        <font-awesome-icon
          v-if="filter.collapsible"
          icon="caret-right"
          :style="iconStyle"
          class="mr-2"
        />
        {{ filter.label }}
        <button
          v-if="canRemove"
          class="remove-button"
          @click.stop="removeFilter"
        >
          <font-awesome-icon icon="times" />
        </button>
      </div>
    </b-card-header>

    <b-collapse
      :id="filter.id"
      :visible="isOpen"
    >
      <b-card-body>
        <b-form-group :description="description">
          <slot />
        </b-form-group>
      </b-card-body>
    </b-collapse>
  </b-card>
</template>

<script>
import Vue from 'vue'

import { library } from '@fortawesome/fontawesome-svg-core'
import { faCaretRight, faTimes } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faCaretRight, faTimes)

export default Vue.extend({
  name: 'FilterCard',
  components: { FontAwesomeIcon },
  props: {
    filter: {
      type: Object,
      default: () => {}
    },
    description: {
      type: String,
      required: false,
      default: () => ''
    },
    canRemove: {
      type: Boolean,
      required: false,
      default: () => false
    }
  },
  computed: {
    iconStyle () {
      return {
        transform: `rotate(${this.filter.collapsed ? 90 : 0}deg)`,
        transition: 'transform 0.2s'
      }
    },
    isOpen: function () {
      if (this.filter.collapsible) {
        return this.filter.collapsed
      }

      return true
    }
  },
  methods: {
    removeFilter () {
      this.filter.active = false
    },
    toggleCollapse: function () {
      if (this.filter.collapsible) {
        this.filter.collapsed = !this.filter.collapsed
      }
    }
  }
})
</script>

<style scoped>
  .card-header.collapsable {
    cursor: pointer;
  }
  .form-group {
    margin-bottom:0;
  }
  .title{
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
  .remove-button{
    background: none;
    border: 0;
    transition: opacity 0.2s, color 0.2s;
    opacity: 0;
    height: inherit;
    width: 1.5em;
    text-align: center;
    display: inline-block;
    position: absolute;
    right: 10px;
    cursor: pointer;
  }
  .remove-button:hover{
    color: var(--danger);
  }
  .filter-card:hover .remove-button{
    opacity: 1;
  }
  .sortable-ghost{
    border-style: dashed;
  }
  .sortable-ghost > div{
    opacity: 0.2;
  }
</style>
