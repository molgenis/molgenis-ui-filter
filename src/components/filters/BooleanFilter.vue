<template>
  <label class="slider-checkbox-container">
    <span
      v-if="falseLabel"
      class="label-text"
    >{{ falseLabel }}</span>
    <div class="slider-checkbox">
      <div class="inner-slider">
        <input
          v-model="selection"
          type="checkbox"
        >
        <div class="slider-knob bg-light" />
        <div
          class="slider-background"
          :class="isSet ? 'bg-primary' : 'bg-secondary'"
        />
      </div>
    </div>
    <span
      v-if="trueLabel"
      class="label-text"
    >{{ trueLabel }}</span>
  </label>
</template>

<script>
export default {
  props: {
    name: {
      type: String,
      required: true
    },
    options: {
      type: [Function],
      required: true
    }
  },
  data () {
    return {
      isSet: Boolean,
      trueLabel: String,
      falseLabel: String
    }
  },
  computed: {
    selection: {
      get () {
        return this.isSet
      },
      set (value) {
        this.$emit('input', value.length === 0 ? undefined : value)
        this.isSet = value
      }
    }
  },
  created () {
    this.options().then(response => {
      this.trueLabel = response.trueLabel
      this.falseLabel = response.falseLabel
    })
  }
}
</script>

<style scoped>
.slider-checkbox-container {
  font-size: 62.5%; /* 1em = 10px */
  display: flex;
  flex-flow: row;
  align-items: center;
}
.slider-checkbox {
  position: relative;
}
.slider-knob {
  width: 1.6em;
  height: 1.6em;
  display: inline-block;
  border-radius: 1em;
  margin: 0 0 0 1px;
  z-index: 10;
  transition: all 0.2s ease;
}
.slider-background {
  position: absolute;
  width: 3.5em;
  height: 100%;
  border-radius: 0.8em;
  transition: all 0.2s ease;
}

input {
  display: none;
}
input[type='checkbox']:checked + .slider-knob {
  /* this is why you want to have 1 rem === 10px; now working with ems*/
  transform: translateX(1.7em);
  transition: all 0.2s ease;
}

.inner-slider {
  cursor: pointer;
  display: flex;
  align-items: center;
  flex-direction: row;
  border-radius: 0.8em;
  margin: 0;
  background-color: lightgrey;
  height: 1.8em;
  width: 3.5em;
}

.label-text {
  font-size: 1.6em;
}

.label-text:first-child {
  margin-right: 1.6em;
}
.label-text:last-child {
  margin-left: 1.6em;
}
</style>
