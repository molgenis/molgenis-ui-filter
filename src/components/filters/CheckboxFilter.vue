<template>
  <div>
    <b-form-checkbox-group
      v-model="filter.selection"
      stacked
      :options="visibleOptions"
      :name="filter.name"
      value-field="id"
      text-field="name"
    />
    <span v-if="bulkOperation">
      <b-link
        v-if="filter.options.length > filter.maxVisibleOptions"
        class="card-link"
        @click="showMoreToggled = !showMoreToggled"
      >
        {{ showMoreText }}
      </b-link>

      <b-link
        v-if="filter.options.length > filter.maxVisibleOptions"
        class="toggle-select card-link"
        @click="toggleSelect"
      >
        {{ filter.selection.length ? 'Deselect all' : 'Select all' }}
      </b-link>
    </span>
  </div>
</template>

<style>
  .card-link { font-style: italic; font-size: small; }
</style>

<script>
export default {
  props: {
    bulkOperation: {
      type: Boolean,
      required: false,
      default: () => true
    },
    filter: {
      required: true,
      type: Object
    }
  },
  data () {
    return {
      showMoreToggled: false
    }
  },
  computed: {
    visibleOptions () {
      if (this.showMoreToggled) {
        return this.filter.options
      }
      return this.filter.options.slice(0, this.filter.maxVisibleOptions)
    },
    showMoreText () {
      if (this.showMoreToggled) {
        return 'Show less'
      }

      return `Show ${this.filter.options.length - this.filter.maxVisibleOptions} more`
    }

  },
  created: async function () {
    this.filter.options = await this.filter.provider()
  },
  methods: {
    toggleSelect () {
      if (this.filter.selection.length) {
        this.filter.selection = []
      } else {
        this.filter.selection = this.visibleOptions.map((f) => f.id)
      }
    }
  }
}
</script>
