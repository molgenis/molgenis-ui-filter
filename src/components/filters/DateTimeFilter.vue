<template>
  <b-input-group class="flex-nowrap">
    <date-range-picker
      ref="picker"
      v-model="dateRange"
      :auto-apply="false"
      class="flex-grow-1"
      :locale-data="{ firstDay: 1, format: 'yyyy-mm-dd HH:MM:ss' }"
      :linked-calendars="false"
      :opens="filter.opens"
      :always-show-calendars="true"
      :ranges="false"
      :show-dropdowns="true"
      :show-week-numbers="false"
      :single-date-picker="!filter.range"
      :time-picker="filter.time"
      :time-picker24hour="true"
      :min-date="filter.min"
      :max-date="filter.max"
    >
      <template v-slot:input="picker">
        {{ dateRange.startDate | formatDate(filter.time) }}
      </template>
    </date-range-picker>

    <b-input-group-append>
      <b-button
        variant="outline-secondary"
        @click="clearDateRange"
      >
        <font-awesome-icon icon="times" />
      </b-button>
    </b-input-group-append>
  </b-input-group>
</template>

<script>
import 'vue2-daterange-picker/dist/vue2-daterange-picker.css'

import Vue from 'vue'
import DateRangePicker from 'vue2-daterange-picker'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faTimes } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'

library.add(faTimes)

const emptyDateRange = {
  startDate: null,
  endDate: null
}

export default Vue.extend({
  name: 'DateTimeFilter',
  components: { DateRangePicker, FontAwesomeIcon },
  filters: {
    formatDate: (date, time) => {
      if (!date) return 'Select...'
      return time ? date.toLocaleString() : date.toLocaleDateString()
    }
  },
  // eslint-disable-next-line vue/require-prop-types
  props: ['filter'],
  data: function () {
    return {
      dateRange: {
        startDate: null,
        endDate: null
      }
    }
  },
  watch: {
    // Convert the selection format of vue2-daterange-picker to
    // be in sync with the preferred filter selection format.
    'dateRange': function toSelection (dateRange) {
      let converted
      if (this.filter.range) {
        converted = [null, null]
        if (dateRange.startDate) {
          converted[0] = this.filter.time ? dateRange.startDate.toLocaleString() : dateRange.startDate.toLocaleDateString()
        }

        if (dateRange.endDate) {
          converted[1] = this.filter.time ? dateRange.endDate.toLocaleString() : dateRange.startDate.toLocaleDateString()
        }
      } else {
        converted = null
        if (dateRange.startDate) {
          converted = this.filter.time ? dateRange.startDate.toLocaleString() : dateRange.startDate.toLocaleDateString()
        }
      }

      this.filter.selection = converted
    }
  },
  methods: {
    clearDateRange: function () {
      this.dateRange = emptyDateRange
      if (this.filter.range) {
        this.filter.selection = [null, null]
      } else {
        this.filter.selection = null
      }
    }
  }
})
</script>
<style lang="css">
  .form-control.reportrange-text {
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
  }
</style>
