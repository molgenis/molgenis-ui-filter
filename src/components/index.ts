export { default as FilterCard } from './FilterCard.vue'
export { default as FilterContainer } from './FilterContainer.vue'
export { default as AddFilterModal } from './AddFilterModal.vue'
export { default as ActiveFilters } from './ActiveFilters.vue'
