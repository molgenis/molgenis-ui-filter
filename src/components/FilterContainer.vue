<template>
  <div
    class="filter-container"
    @mouseup="drag=false"
  >
    <b-collapse
      id="mobile-button-toggle"
      :visible="doCollapse"
    >
      <button
        class="btn w-100 my-2 btn-outline-secondary"
        @click="mobileToggle=!mobileToggle"
      >
        {{ mobileToggle?'Hide filters':'Show filters' }}
      </button>
    </b-collapse>

    <b-collapse
      id="mobile-toggle"
      :visible="!doCollapse || mobileToggle"
    >
      <draggable
        v-model="filters"
        :disabled="!doDragDrop"
        :class="{'dragdrop': doDragDrop, 'dragging': drag}"
        @choose="drag=true"
        @end="drag=false"
      >
        <transition-group>
          <filter-card
            v-for="filter in visibleFilters"
            :key="filter.id"
            :filter="filter"
            :can-remove="canEdit"
            @removeFilter="filter.visible = false"
          >
            <component
              :is="filter.type"
              :filter="filter"
            />
          </filter-card>
        </transition-group>
      </draggable>

      <add-filter-modal
        v-if="canEdit && invisibleFilters.length > 0"
        v-model="filters.visible"
        :filters="invisibleFilters"
      />
    </b-collapse>
  </div>
</template>

<script>
import AddFilterModal from './AddFilterModal.vue'
import { FilterCard } from '.'
import { StringFilter, CheckboxFilter, NumberFilter, RangeFilter, MultiFilter } from './filters/'
import draggable from 'vuedraggable'

export default {
  name: 'FilterContainer',
  components: { AddFilterModal, StringFilter, CheckboxFilter, FilterCard, NumberFilter, RangeFilter, MultiFilter, draggable },
  props: {
    canEdit: {
      type: Boolean,
      required: false,
      default: () => false
    }
  },
  data () {
    return {
      drag: false,
      width: 0,
      mobileToggle: false
    }
  },
  computed: {
    doCollapse () {
      // Bootstrap's mobile collapse width
      return this.width <= 576
    },
    doDragDrop () {
      return this.canEdit && !this.doCollapse
    },
    invisibleFilters () {
      return this.filters.filter(filter => !filter.visible)
    },
    visibleFilters () {
      return this.filters.filter(filter => filter.visible)
    }
  },
  created () {
    window.addEventListener('resize', this.handleResize)
    this.handleResize()
  },
  destroyed () {
    window.removeEventListener('resize', this.handleResize)
  },
  methods: {
    handleResize () {
      this.width = window.innerWidth
    }
  },
  store: ['filters']
}
</script>

<style scoped>
  .dragdrop{
    cursor: grab;
  }
  .dragdrop.dragging{
    cursor: grabbing;
  }
</style>
