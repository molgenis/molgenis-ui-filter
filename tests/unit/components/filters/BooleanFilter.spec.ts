import { mount } from '@vue/test-utils'
import { BooleanFilter } from '@/components/filters/'

describe('BooleanFilter.vue', () => {
  const optionsPromise = () => {
    return new Promise(
      function (resolve) {
        resolve({ trueLabel: 'Public', falseLabel: 'Private' })
      }
    )
  }

  const propsData = {
    name: 'boolean-filter',
    label: 'Boolean filter',
    options: optionsPromise,
    type: 'boolean-filter'
  }

  const wrapper = mount(BooleanFilter, { propsData })
  wrapper.vm.$emit = jest.fn()

  it('matches the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
