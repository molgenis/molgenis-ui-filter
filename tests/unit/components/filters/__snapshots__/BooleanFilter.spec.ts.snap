// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooleanFilter.vue matches the snapshot 1`] = `
<label
  class="slider-checkbox-container"
  label="Boolean filter"
  type="boolean-filter"
>
  <span
    class="label-text"
  >
    Private
  </span>
   
  <div
    class="slider-checkbox"
  >
    <div
      class="inner-slider"
    >
      <input
        type="checkbox"
      />
       
      <div
        class="slider-knob bg-light"
      />
       
      <div
        class="slider-background bg-primary"
      />
    </div>
  </div>
   
  <span
    class="label-text"
  >
    Public
  </span>
</label>
`;
